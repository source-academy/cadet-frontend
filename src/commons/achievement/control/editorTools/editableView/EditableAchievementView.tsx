import { Button, Card, Dialog } from '@blueprintjs/core';
import React, { useState } from 'react';

import { AchievementView } from '../../../../../features/achievement/AchievementTypes';
import EditableViewDescription from './EditableViewDescription';
import EditableViewImage from './EditableViewImage';
import EditableViewText from './EditableViewText';

type EditableAchievementViewProps = {
  title: string;
  view: AchievementView;
  changeView: any;
};

function EditableAchievementView(props: EditableAchievementViewProps) {
  const { title, view, changeView } = props;

  const [isDialogOpen, setDialogOpen] = useState<boolean>(false);

  const setDescription = (newDescription: string) => {
    const newView = {
      description: newDescription,
      canvasUrl: view.canvasUrl,
      completionText: view.completionText
    };
    changeView(newView);
  };

  const setCompletionText = (newCompletionText: string) => {
    const newView = {
      description: view.description,
      canvasUrl: view.canvasUrl,
      completionText: newCompletionText
    };
    changeView(newView);
  };

  const setCanvasUrl = (newCanvasUrl: string) => {
    const newView = {
      description: view.description,
      canvasUrl: newCanvasUrl,
      completionText: view.completionText
    };
    changeView(newView);
  };

  return (
    <div>
      <div>
        <Button text={'Edit View'} onClick={() => setDialogOpen(!isDialogOpen)} />
      </div>
      <Dialog
        onClose={() => setDialogOpen(!isDialogOpen)}
        isOpen={isDialogOpen}
        title={'Edit View'}
        usePortal={false}
      >
        <div className="view-editor">
          <Card className="background-card">
            <h1>{title} </h1>

            <EditableViewImage
              canvasUrl={view.canvasUrl}
              title={title}
              setCanvasUrl={setCanvasUrl}
            />
            <EditableViewDescription
              description={view.description}
              setDescription={setDescription}
            />
            <EditableViewText goalText={view.completionText} setGoalText={setCompletionText} />
          </Card>
        </div>
      </Dialog>
    </div>
  );
}

export default EditableAchievementView;
