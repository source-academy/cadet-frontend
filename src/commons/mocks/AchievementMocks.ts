import {
  AchievementAbility,
  AchievementGoal,
  AchievementItem,
  GoalType
} from '../../features/achievement/AchievementTypes';

export const mockAchievements: AchievementItem[] = [
  {
    id: 0,
    title: 'Rune Master',
    ability: AchievementAbility.CORE,
    isTask: true,
    position: 1,
    prerequisiteIds: [2, 1],
    goalIds: [0],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/rune-master-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Rune Master achievement'
    }
  },
  {
    id: 1,
    title: 'Beyond the Second Dimension',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 6, 12, 30, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [1],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/btsd-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Beyond the Second Dimension achievement'
    }
  },
  {
    id: 2,
    title: 'Colorful Carpet',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 8, 9, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [2],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/colorful-carpet-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Colorful Carpet achievement'
    }
  },
  {
    id: 3,
    title: '',
    ability: AchievementAbility.CORE,
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [],
    cardTileUrl: '',
    view: {
      canvasUrl: '',
      description: '',
      completionText: ''
    }
  },
  {
    id: 4,
    title: 'Curve Wizard',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 8, 15, 0, 0, 0),
    isTask: true,
    position: 4,
    prerequisiteIds: [5, 6],
    goalIds: [3],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-wizard-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Wizard achievement'
    }
  },
  {
    id: 5,
    title: 'Curve Introduction',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 28, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [4],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-introduction-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Introduction achievement'
    }
  },
  {
    id: 6,
    title: 'Curve Manipulation',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 8, 5, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [5],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-manipulation-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Manipulation achievement'
    }
  },
  {
    id: 7,
    title: 'The Source-rer',
    ability: AchievementAbility.EFFORT,
    deadline: new Date(2020, 7, 21, 0, 0, 0),
    isTask: true,
    position: 3,
    prerequisiteIds: [],
    goalIds: [6, 7],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/the-source-rer-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed The Source-rer achievement'
    }
  },
  {
    id: 8,
    title: 'Power of Friendship',
    ability: AchievementAbility.COMMUNITY,
    isTask: true,
    position: 2,
    prerequisiteIds: [9],
    goalIds: [],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/power-of-friendship-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Power of Friendship achievement'
    }
  },
  {
    id: 9,
    title: 'Piazza Guru',
    ability: AchievementAbility.COMMUNITY,
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [8],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/piazza-guru-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Piazza Guru achievement'
    }
  },
  {
    id: 10,
    title: "That's the Spirit",
    ability: AchievementAbility.EXPLORATION,
    isTask: true,
    position: 5,
    prerequisiteIds: [],
    goalIds: [9],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/thats-the-spirit-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: "Congratulations! You have completed That's the Spirit achievement"
    }
  },
  {
    id: 11,
    title: 'Kool Kidz',
    ability: AchievementAbility.FLEX,
    isTask: true,
    position: 6,
    prerequisiteIds: [],
    goalIds: [],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/annotated-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description: 'Diz for teh K00L K1dz',
      completionText: 'U SO G00D'
    }
  }
];

export const mockGoals: AchievementGoal[] = [
  {
    id: 0,
    text: 'Bonus for completing Colorful Carpet & Beyond the Second Dimension Achievements',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `AND(
        {
          event: 'achievement',
          restriction: 2,
        },
        {
          event: 'achievement',
          restriction: 1,
        },
      )`
    },
    exp: 0,
    completed: false
  },
  {
    id: 1,
    text: 'XP earned from Beyond the Second Dimension Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 5, requiredCompletionExp: 200 },
    exp: 213,
    completed: true
  },
  {
    id: 2,
    text: 'XP earned from Colorful Carpet Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 3, requiredCompletionExp: 200 },
    exp: 0,
    completed: false
  },
  {
    id: 3,
    text: 'Bonus for completing Curve Introduction & Curve Manipulation Achievements',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `AND(
      {
        event: 'achievement',
        restriction: 4,
      },
      {
        event: 'achievement',
        restriction: 5,
      },
    )`
    },
    exp: 0,
    completed: false
  },
  {
    id: 4,
    text: 'XP earned from Curve Introduction Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 7, requiredCompletionExp: 150 },
    exp: 178,
    completed: true
  },
  {
    id: 5,
    text: 'XP earned from Curve Manipulation Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 8, requiredCompletionExp: 200 },
    exp: 191,
    completed: false
  },
  {
    id: 6,
    text: 'Submit Source 3 Path',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `
      {
        event: 'assessment-submission',
        restriction: paths.SOURCE_THREE,
      },
    `
    },
    exp: 100,
    completed: true
  },
  {
    id: 7,
    text: 'XP earned from Source 3 Path',
    maxExp: 300,
    type: GoalType.ASSESSMENT,
    meta: {
      assessmentId: 12,
      requiredCompletionExp: 300
    },
    exp: 300,
    completed: true
  },
  {
    id: 8,
    text: 'Each Top Voted answer in Piazza gives 10 XP',
    maxExp: 100,
    type: GoalType.MANUAL,
    meta: {},
    exp: 40,
    completed: false
  },
  {
    id: 9,
    text: 'Submit 1 PR to Source Academy Github',
    maxExp: 100,
    type: GoalType.MANUAL,
    meta: {},
    exp: 100,
    completed: true
  },
  {
    id: 10,
    text: 'Be the Koolest Kidz in SOC by redeeming this 100 XP achievement yourself',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `
      {
        event: 'achievement',
        restriction: 'false',
      },
    `
    },
    exp: 0,
    completed: false
  }
];
