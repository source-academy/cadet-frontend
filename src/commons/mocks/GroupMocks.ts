import { GradingSummary } from '../../features/dashboard/DashboardTypes';

export const mockGradingSummary: GradingSummary = [
  {
    leaderName: 'John',
    groupName: 'Mock Group 1',
    ungradedMissions: 123,
    submittedMissions: 200,
    ungradedSidequests: 100,
    submittedSidequests: 117
  },
  {
    leaderName: 'Molly',
    groupName: 'Mock Group 2',
    ungradedMissions: 1232,
    submittedMissions: 205430,
    ungradedSidequests: 345,
    submittedSidequests: 11547
  },
  {
    leaderName: 'Lenny',
    groupName: 'Mock Group 3',
    ungradedMissions: 1532,
    submittedMissions: 22200,
    ungradedSidequests: 134500,
    submittedSidequests: 6777
  }
];
