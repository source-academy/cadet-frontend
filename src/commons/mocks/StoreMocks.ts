import { Store } from 'redux';
import mockStore from 'redux-mock-store';

import {
  defaultAcademy,
  defaultApplication,
  defaultDashboard,
  defaultPlayground,
  defaultSession,
  defaultWorkspaceManager,
  OverallState
} from '../application/ApplicationTypes';

export function mockInitialStore<P>(): Store<OverallState> {
  const createStore = (mockStore as any)();
  const state: OverallState = {
    academy: defaultAcademy,
    application: defaultApplication,
    dashboard: defaultDashboard,
    playground: defaultPlayground,
    workspaces: defaultWorkspaceManager,
    session: defaultSession
  };
  return createStore(state);
}
