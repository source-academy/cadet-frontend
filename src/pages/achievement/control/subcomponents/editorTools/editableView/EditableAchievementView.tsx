import { Button, Card, Dialog } from '@blueprintjs/core';
import React, { useState } from 'react';

import { AchievementViewItem } from '../../../../../../features/achievement/AchievementTypes';
import EditableViewDescription from './EditableViewDescription';
import EditableViewImage from './EditableViewImage';
import EditableViewText from './EditableViewText';

type EditableAchievementViewProps = {
  title: string;
  view: AchievementViewItem;
  changeView: any;
};

function EditableAchievementView(props: EditableAchievementViewProps) {
  const { title, view, changeView } = props;

  const [isDialogOpen, setDialogOpen] = useState<boolean>(false);

  const [viewData, setViewData] = useState<AchievementViewItem>(view);

  const { canvasUrl, description, completionText } = viewData;

  const setDescription = (description: string) => {
    setViewData({
      ...viewData,
      description: description
    });
    changeView(viewData);
  };

  const setCompletionText = (completionText: string) => {
    setViewData({
      ...viewData,
      completionText: completionText
    });
    changeView(viewData);
  };

  const setcanvasUrl = (canvasUrl: string) => {
    setViewData({
      ...viewData,
      canvasUrl: canvasUrl
    });
    changeView(viewData);
  };

  return (
    <div>
      <div>
        <Button text={'Edit View'} onClick={() => setDialogOpen(!isDialogOpen)} />
      </div>
      <Dialog
        onClose={() => setDialogOpen(!isDialogOpen)}
        isOpen={isDialogOpen}
        title={'Edit View'}
        usePortal={false}
      >
        <div className="view-editor">
          <Card className="background-card">
            <h1>{title} </h1>

            <EditableViewImage canvasUrl={canvasUrl} title={title} setcanvasUrl={setcanvasUrl} />

            <EditableViewDescription description={description} setDescription={setDescription} />
            <EditableViewText goalText={completionText} setGoalText={setCompletionText} />
          </Card>
        </div>
      </Dialog>
    </div>
  );
}

export default EditableAchievementView;
