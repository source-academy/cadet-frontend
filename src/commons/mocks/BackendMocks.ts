import { SagaIterator } from 'redux-saga';
import { call, put, select, takeEvery } from 'redux-saga/effects';

import { FETCH_GROUP_GRADING_SUMMARY } from '../../features/dashboard/DashboardTypes';
import { Grading, GradingOverview, GradingQuestion } from '../../features/grading/GradingTypes';
import { store } from '../../pages/createStore';
import { GameState, OverallState, Role } from '../application/ApplicationTypes';
import {
  ACKNOWLEDGE_NOTIFICATIONS,
  FETCH_ASSESSMENT,
  FETCH_AUTH,
  FETCH_GRADING,
  FETCH_GRADING_OVERVIEWS,
  FETCH_NOTIFICATIONS,
  SUBMIT_ANSWER,
  SUBMIT_GRADING,
  SUBMIT_GRADING_AND_CONTINUE,
  UNSUBMIT_SUBMISSION
} from '../application/types/SessionTypes';
import { FETCH_ASSESSMENT_OVERVIEWS, Question } from '../assessment/AssessmentTypes';
import {
  Notification,
  NotificationFilterFunction
} from '../notificationBadge/NotificationBadgeTypes';
import { actions } from '../utils/ActionsHelper';
import { history } from '../utils/HistoryHelper';
import { showSuccessMessage, showWarningMessage } from '../utils/NotificationsHelper';
import { WorkspaceLocation } from '../workspace/WorkspaceTypes';
import { mockAssessmentOverviews, mockAssessments } from './AssessmentMocks';
import { mockFetchGrading, mockFetchGradingOverview } from './GradingMocks';
import { mockGradingSummary } from './GroupMocks';
import { mockNotifications } from './UserMocks';

export function* mockBackendSaga(): SagaIterator {
  yield takeEvery(FETCH_AUTH, function* (action: ReturnType<typeof actions.fetchAuth>) {
    const tokens = {
      accessToken: 'accessToken',
      refreshToken: 'refreshToken'
    };
    const user = {
      userId: 123,
      name: 'DevStaff',
      role: 'staff' as Role,
      group: '1F',
      story: {
        story: 'mission-1',
        playStory: true
      },
      grade: 0,
      gameState: {} as GameState
    };
    store.dispatch(actions.setTokens(tokens));
    store.dispatch(actions.setUser(user));
    yield history.push('/academy');
  });

  yield takeEvery(FETCH_ASSESSMENT_OVERVIEWS, function* () {
    yield put(actions.updateAssessmentOverviews([...mockAssessmentOverviews]));
  });

  yield takeEvery(FETCH_ASSESSMENT, function* (action: ReturnType<typeof actions.fetchAssessment>) {
    const id = action.payload;
    const assessment = mockAssessments[id - 1];
    yield put(actions.updateAssessment({ ...assessment }));
  });

  yield takeEvery(
    FETCH_GRADING_OVERVIEWS,
    function* (action: ReturnType<typeof actions.fetchGradingOverviews>) {
      const accessToken = yield select((state: OverallState) => state.session.accessToken);
      const filterToGroup = action.payload;
      const gradingOverviews = yield call(() =>
        mockFetchGradingOverview(accessToken, filterToGroup)
      );
      if (gradingOverviews !== null) {
        yield put(actions.updateGradingOverviews([...gradingOverviews]));
      }
    }
  );

  yield takeEvery(FETCH_GRADING, function* (action: ReturnType<typeof actions.fetchGrading>) {
    const submissionId = action.payload;
    const accessToken = yield select((state: OverallState) => state.session.accessToken);
    const grading = yield call(() => mockFetchGrading(accessToken, submissionId));
    if (grading !== null) {
      yield put(actions.updateGrading(submissionId, [...grading]));
    }
  });

  yield takeEvery(SUBMIT_ANSWER, function* (action: ReturnType<typeof actions.submitAnswer>) {
    const questionId = action.payload.id;
    const answer = action.payload.answer;
    // Now, update the answer for the question in the assessment in the store
    const assessmentId = yield select(
      (state: OverallState) => state.workspaces.assessment.currentAssessment!
    );
    const assessment = yield select((state: OverallState) =>
      state.session.assessments.get(assessmentId)
    );
    const newQuestions = assessment.questions.slice().map((question: Question) => {
      if (question.id === questionId) {
        question.answer = answer;
      }
      return question;
    });
    const newAssessment = {
      ...assessment,
      questions: newQuestions
    };
    yield put(actions.updateAssessment(newAssessment));
    yield call(showSuccessMessage, 'Saved!', 1000);
    return yield put(actions.updateHasUnsavedChanges('assessment' as WorkspaceLocation, false));
  });

  yield takeEvery(
    UNSUBMIT_SUBMISSION,
    function* (action: ReturnType<typeof actions.unsubmitSubmission>) {
      const { submissionId } = action.payload;
      const overviews: GradingOverview[] = yield select(
        (state: OverallState) => state.session.gradingOverviews || []
      );
      const index = overviews.findIndex(
        overview =>
          overview.submissionId === submissionId && overview.submissionStatus === 'submitted'
      );
      if (index === -1) {
        yield call(showWarningMessage, '400: Bad Request');
        return;
      }
      const newOverviews = (overviews as GradingOverview[]).map(overview => {
        if (overview.submissionId === submissionId) {
          return { ...overview, submissionStatus: 'attempted' };
        }
        return overview;
      });
      yield call(showSuccessMessage, 'Unsubmitted!', 1000);
      yield put(actions.updateGradingOverviews(newOverviews));
    }
  );

  const sendGrade = function* (
    action: ReturnType<typeof actions.submitGrading | typeof actions.submitGradingAndContinue>
  ) {
    const { submissionId, questionId, gradeAdjustment, xpAdjustment, comments } = action.payload;
    // Now, update the grade for the question in the Grading in the store
    const grading: Grading = yield select((state: OverallState) =>
      state.session.gradings.get(submissionId)
    );
    const newGrading = grading.slice().map((gradingQuestion: GradingQuestion) => {
      if (gradingQuestion.question.id === questionId) {
        gradingQuestion.grade = {
          gradeAdjustment,
          xpAdjustment,
          roomId: gradingQuestion.grade.roomId,
          grade: gradingQuestion.grade.grade,
          xp: gradingQuestion.grade.xp,
          comments
        };
      }
      return gradingQuestion;
    });
    yield put(actions.updateGrading(submissionId, newGrading));
    yield call(showSuccessMessage, 'Submitted!', 1000);
  };

  const sendGradeAndContinue = function* (
    action: ReturnType<typeof actions.submitGradingAndContinue>
  ) {
    const { submissionId } = action.payload;
    yield* sendGrade(action);

    const currentQuestion = yield select(
      (state: OverallState) => state.workspaces.grading.currentQuestion
    );
    /**
     * Move to next question for grading: this only works because the
     * SUBMIT_GRADING_AND_CONTINUE Redux action is currently only
     * used in the Grading Workspace
     *
     * If the questionId is out of bounds, the componentDidUpdate callback of
     * GradingWorkspace will cause a redirect back to '/academy/grading'
     */
    yield history.push(`/academy/grading/${submissionId}/${(currentQuestion || 0) + 1}`);
  };

  yield takeEvery(SUBMIT_GRADING, sendGrade);

  yield takeEvery(SUBMIT_GRADING_AND_CONTINUE, sendGradeAndContinue);

  yield takeEvery(
    ACKNOWLEDGE_NOTIFICATIONS,
    function* (action: ReturnType<typeof actions.acknowledgeNotifications>) {
      const notificationFilter: NotificationFilterFunction | undefined = action.payload.withFilter;

      const notifications: Notification[] = yield select(
        (state: OverallState) => state.session.notifications
      );

      let notificationsToAcknowledge = notifications;

      if (notificationFilter) {
        notificationsToAcknowledge = notificationFilter(notifications);
      }

      if (notificationsToAcknowledge.length === 0) {
        return;
      }

      const ids = notificationsToAcknowledge.map(n => n.id);

      const newNotifications: Notification[] = notifications.filter(
        notification => !ids.includes(notification.id)
      );

      yield put(actions.updateNotifications(newNotifications));
    }
  );

  yield takeEvery(
    FETCH_NOTIFICATIONS,
    function* (action: ReturnType<typeof actions.fetchNotifications>) {
      yield put(actions.updateNotifications(mockNotifications));
    }
  );

  yield takeEvery(FETCH_GROUP_GRADING_SUMMARY, function* () {
    yield put(actions.updateGroupGradingSummary([...mockGradingSummary]));
  });
}
