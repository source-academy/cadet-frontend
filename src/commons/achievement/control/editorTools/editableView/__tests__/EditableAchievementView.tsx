import { mount } from 'enzyme';
import React from 'react';

import { mockAchievements } from '../../../../../mocks/AchievementMocks';
import EditableAchievementView from '../EditableAchievementView';

const mockProps = {
  title: 'Sample Title',
  view: mockAchievements[0].view,
  changeView: () => {}
};

test('EditableAchievementView component renders correctly', () => {
  const view = <EditableAchievementView {...mockProps} />;
  const tree = mount(view);
  expect(tree.debug()).toMatchSnapshot();
});
