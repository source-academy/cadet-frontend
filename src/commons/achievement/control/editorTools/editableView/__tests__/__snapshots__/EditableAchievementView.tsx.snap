// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableAchievementView component renders correctly 1`] = `
"<EditableAchievementView title=\\"Sample Title\\" view={{...}} changeView={[Function: changeView]}>
  <div>
    <div>
      <Blueprint3.Button text=\\"Edit View\\" onClick={[Function: onClick]}>
        <button type=\\"button\\" onClick={[Function: onClick]} className=\\"bp3-button\\" disabled={[undefined]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
          <Blueprint3.Icon icon={[undefined]} />
          <span className=\\"bp3-button-text\\">
            Edit View
          </span>
          <Blueprint3.Icon icon={[undefined]} />
        </button>
      </Blueprint3.Button>
    </div>
    <Blueprint3.Dialog onClose={[Function: onClose]} isOpen={false} title=\\"Edit View\\" usePortal={false} canOutsideClickClose={true}>
      <Blueprint3.Overlay onClose={[Function: onClose]} isOpen={false} title=\\"Edit View\\" usePortal={false} canOutsideClickClose={true} className=\\"bp3-overlay-scroll-container\\" hasBackdrop={true} autoFocus={true} backdropProps={{...}} canEscapeKeyClose={true} enforceFocus={true} lazy={true} transitionDuration={300} transitionName=\\"bp3-overlay\\" />
    </Blueprint3.Dialog>
  </div>
</EditableAchievementView>"
`;
